sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala

```diff
@@ -200,6 +201,43 @@ class V2ExpressionBuilder(
       } else {
         None
       }
+    case substring: Substring =>
+      val childrenExpressions = substring.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == substring.children.length) {
+        Some(new GeneralScalarExpression("SUBSTRING",
+          childrenExpressions.toArray[V2Expression]))
+      } else {
+        None
+      }
+    case Upper(child) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("UPPER", Array[V2Expression](v)))
+    case Lower(child) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("LOWER", Array[V2Expression](v)))
+    case translate: StringTranslate =>
+      val childrenExpressions = translate.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == translate.children.length) {
+        Some(new GeneralScalarExpression("TRANSLATE",
+          childrenExpressions.toArray[V2Expression]))
+      } else {
+        None
+      }
+    case trim: String2TrimExpression =>
+      val childrenExpressions = trim.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == trim.children.length) {
+        Some(new GeneralScalarExpression("TRIM",
+          LiteralValue(UTF8String.fromString(trim.direction), VarcharType(0)) +:
```

Comment:
`StringType` ?

Reply:
StringType will add single quotes to LEADING into 'LEADING'.
And the query result is wrong.

Reply:
cc @cloud-fan